active/0051-strategies.md

```diff
@@ -0,0 +1,441 @@
+
+  toTree(inputTree) {
+    return concat(inputTree, this.options);
+  }
+}
+```
+
+Where `this.options` mimics [`broccoli-concat`
+options](https://github.com/broccolijs/broccoli-concat#advanced-usage).
+
+### Example of possible strategies
+
+Below are some of the strategies that could be implemented.
+
+#### Assets Split Strategy
+
+One of the techniques for improving site speed is isolating changes throughout
+application deployments. Assuming  the application assets are uploaded to CDN,
+the reasoning is very simple: if `ember.js` or `jQuery` (possibly along with
+other third party libraries) don't change with every deployment, why bust CDN
+cache for them?
+
+This strategy could be provided out of the box by Ember CLI as a part of site
+speed improvement story.
+
+```javascript
+class AssetsSplitStrategy {
+  constructor(options) {
+    this.options = options;
+  }
+
+  toTree(inputTree) {
+    return concat(inputTree, {
+      headerFiles: [
+        jQueryPath,
+        emberPath
+      ],
+      outputFile: 'vendor-static.js'
+    });
+  }
+}
+```
+
+#### Debug/Analyse Strategy
+
+Yet another application of strategies would be to run different analysis on the
+Ember applications. Take
+[broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser),
+for example. This could be a strategy and easily incorporated into the build.
+
+#### Engine Strategy
+
+Introducing `Strategy` as a concept allows us to re-think the way we approach
+Ember Engines as well. We can start thinking about engines as a strategy of
+certain type. In fact, it's very similar to `ConcatenationStrategy` described
+above.
+
++ reduces the API surface (as opposed to `ember-engines` and
+  `ember-asset-loader`, only one strategy)
++ makes things explicit, meaning that you have to register a strategy
+
+## Registering custom strategies
+
+Ember applications would be able to register custom strategies via
+`ember-cli-build.js` file by passing a `strategies` property to the constructor.
+
+Passing `strategies` as an option to the constructor will clobber default
+strategies that Ember CLI provides out of the box (they mimic existing behaviour
+of creating `dist` folder with final assets).
+
+Ember CLI will provide a way to use default strategies that people can
+take advantage of.
+
+For example:
+
+```javascript
+// ember-cli-build.js
+const {
+  createDefaultApplicationStrategy,
+  createDefaultVendorStrategy
+} = require('ember-cli/lib/broccoli/strategies');
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+// no operation strategy that puts app/add-ons tree inside of
+// `noop-output` folder
+class Noop {
+  constructor(options) {
+    this.options = options;
+  }
+
+  toTree(inputTree) {
+    return inputTree;
+  }
+}
+
+module.exports = function(defaults) {
+  const app = new EmberApp(defaults, { });
+
+  app.strategies = [
+    new Noop(),
+    createDefaultApplicationStrategy(app)
+  ].concat(createDefaultVendorStrategy(app));
```

Comment:
Why does `createDefaultVendorStrategy` return an array of strategies?

Reply:
it's mostly have to do with the fact that vendor can produce several files, not just one.